packages/base-controller/src/ControllerMessenger.ts

```diff
@@ -1,38 +1,58 @@
-export type ActionHandler<Action, ActionType> = (
+export type ActionHandler<
+  Action extends ActionConstraint,
+  ActionType = Action['type'],
+> = (
   ...args: ExtractActionParameters<Action, ActionType>
 ) => ExtractActionResponse<Action, ActionType>;
-export type ExtractActionParameters<Action, T> = Action extends {
+
+export type ExtractActionParameters<
+  Action extends ActionConstraint,
+  T = Action['type'],
+> = Action extends {
   type: T;
-  handler: (...args: infer H) => any;
+  handler: (...args: infer H) => unknown;
 }
   ? H
   : never;
-export type ExtractActionResponse<Action, T> = Action extends {
+
+export type ExtractActionResponse<
+  Action extends ActionConstraint,
+  T = Action['type'],
+> = Action extends {
   type: T;
-  handler: (...args: any) => infer H;
+  handler: (...args: never[]) => infer HandlerReturnValue;
 }
-  ? H
+  ? HandlerReturnValue
   : never;
 
-export type ExtractEventHandler<Event, T> = Event extends {
+export type ExtractEventHandler<
+  Event extends EventConstraint,
+  T = Event['type'],
+> = Event extends {
   type: T;
   payload: infer P;
 }
   ? P extends unknown[]
     ? (...payload: P) => void
     : never
   : never;
-export type ExtractEventPayload<Event, T> = Event extends {
+
+export type ExtractEventPayload<
+  Event extends EventConstraint,
+  T = Event['type'],
+> = Event extends {
   type: T;
   payload: infer P;
 }
-  ? P
+  ? P extends unknown[]
+    ? P
+    : never
   : never;
 
 export type GenericEventHandler = (...args: unknown[]) => void;
 
-export type SelectorFunction<Args extends unknown[], ReturnValue> = (
-  ...args: Args
+export type SelectorFunction<Event extends EventConstraint, ReturnValue> = (
+  ...args: ExtractEventPayload<Event>
```

Comment:
Highlighting for closer review

Reply:
Good catch, that does seem like a much more appropriate type